includes/admin/wc-admin-functions.php

```diff
@@ -206,17 +206,17 @@ function wc_maybe_adjust_line_item_product_stock( $item, $item_quantity = -1 ) {
 		return false;
 	}
 
-	$product               = $item->get_product();
-	$item_quantity         = wc_stock_amount( $item_quantity >= 0 ? $item_quantity : $item->get_quantity() );
-	$already_reduced_stock = wc_stock_amount( $item->get_meta( '_reduced_stock', true ) );
+	$product                = $item->get_product();
+	$item_quantity          = wc_stock_amount( $item_quantity >= 0 ? $item_quantity : $item->get_quantity() );
+	$already_reduced_stock  = wc_stock_amount( $item->get_meta( '_reduced_stock', true ) );
+	$restock_refunded_items = wc_stock_amount( $item->get_meta( '_restock_refunded_items', true ) );
```

Comment:
For existing orders with refunds, `_restock_refunded_items` won't be set, for those old orders, if someone clicks on the update, we will end up increasing the back any reduced stock on those old orders. 

Perhaps we can assume the value of  `$item_quantity - $already_reduced_stock` in case `_restock_refunded_items` meta is not set at all? (And set it to 0- when we update an order, if not already set).

Reply:
Unit tests already exists for this method `wc_maybe_adjust_line_item_product_stock` and it is passing.

Reply:
Ah ok, but as you can see those tests were not enough to catch this bug. Should we still add some more?

Reply:
I am not sure if we can do anything for older orders. Because if `$restock_refunded_items` is not set, it is returning 0 already. So the logic already becomes `$item_quantity - $already_reduced_stock` as you can see here https://github.com/woocommerce/woocommerce/pull/29762/files#diff-0c9a45b990699b400112d411b46f124a4b420725af90363a568f85a7f44b98cbR219

Reply:
Additional unit tests added.

Reply:
> Because if $restock_refunded_items is not set, it is returning 0 already. So the logic already becomes $item_quantity - $already_reduced_stock as you can see here

My bad, I meant initializing the $restock_refunded_itmes to `$refunded_item_quantity`. It will still be a bug, but a bit better I think.

Currently if you:
1. Refund an order with restocking in the older WooCommerce version (or trunk). See that stocks are increased for the product.
2. Click on update, nothing will happen.
3. Switch to a WooCommerce version with this fix, and click on update again. See that stock will be incorrectly reduced for the product. This is because `$restock_refunded_itmes` will return 0.

If we instead change `$restock_refunded_itmes` to be initialized to `refunded_item_quantity` in case its not present, it will be only as bad as the current (trunk) behavior, which is buggy for older orders only when stock is refunded without restocking.

Reply:
I am not sure if that would work correctly because while it would work for older orders that didn't have `$restock_refunded_items` but for new orders it is possible this is also not set yet. Then the same issue continues to happen for new orders.

Reply:
One thought I had is we handle both scenarios differently depending on the `order version`. If this order is from WC `<5.4`, we handle it the old way and if it is `>=5.4`, we handle it with the new logic?

Reply:
>  If this order is from WC <5.4, we handle it the old way and if it is >=5.4, we handle it with the new logic

Ok, sounds good although it does look a bit hacky. I guess there is no better way? Can we set  `restock_refunded_items` to 0 for every new order whenever the first refund happens so that it's not undefined for new orders?

Reply:
So it won't be undefined because we're wrapping it in `wc_stock_amount()` which evaluate it to `0` if it is empty. I have pushed up the changes per our discussion. I have also updated the test to work with WC 5.4+ only. The only thing new here is when you test this PR, you have to set https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-woocommerce.php#L27 to version `5.4`